core/store/src/utils/shard_size.rs

```diff
@@ -0,0 +1,142 @@
+use crate::trie::AccessOptions;
+use crate::trie::iterator::DiskTrieIteratorInner;
+use crate::trie::mem::iter::MemTrieIteratorInner;
+use crate::trie::ops::interface::{GenericTrieInternalStorage, GenericTrieNode};
+use crate::{NibbleSlice, Trie, TrieUpdate, get, set};
+use near_primitives::errors::StorageError;
+use near_primitives::shard_size_contribution::ShardSizeContribution;
+use near_primitives::trie_key::TrieKey;
+use near_primitives::types::AccountId;
+use std::str::FromStr;
+
+/// Add the given shard size contribution to the total amount stored for
+/// the given account ID and all its prefixes.
+pub fn update_contribution(
+    state_update: &mut TrieUpdate,
+    account_id: &AccountId,
+    contribution: ShardSizeContribution,
+) -> Result<(), StorageError> {
+    for i in 0..=account_id.len() {
+        let prefix = &account_id.as_bytes()[..i];
+        let trie_key = TrieKey::ShardSizeContribution { account_id_prefix: prefix.to_vec() };
+        let mut account_contribution: ShardSizeContribution =
+            get(state_update, &trie_key)?.unwrap_or_default();
+        account_contribution += contribution;
+        set(state_update, trie_key, &account_contribution);
+    }
+    Ok(())
+}
+
+/// Find an account ID that splits the shard into two possibly equal parts, according
+/// to the shard size contribution.
+/// Returns `None` if the trie is empty or an empty node is found.
+pub fn find_split_account(trie: &Trie) -> Result<Option<AccountId>, StorageError> {
+    // Find the split key
+    let Some(key) = (match trie.read_lock_memtries() {
+        Some(memtries) => find_split_key(MemTrieIteratorInner::new(&*memtries, trie))?,
+        None => find_split_key(DiskTrieIteratorInner::new(trie))?,
+    }) else {
+        return Ok(None);
+    };
+
+    // Convert the key to account ID by finding the longest prefix that is a valid ID
+    for i in 0..key.len() {
+        let prefix = std::str::from_utf8(&key[..(key.len() - i)]).expect("trie key is non-ASCII");
+        if let Ok(account_id) = AccountId::from_str(prefix) {
+            return Ok(Some(account_id));
+        }
+    }
+    Ok(None)
+}
+
+fn find_split_key<NodePtr, ValueRef, TrieStorage>(
+    trie_storage: TrieStorage,
+) -> Result<Option<Vec<u8>>, StorageError>
+where
+    NodePtr: Copy,
+    TrieStorage: GenericTrieInternalStorage<NodePtr, ValueRef>,
+{
+    let mut key_nibbles = Vec::with_capacity(AccountId::MAX_LEN * 2); // x2 because nibbles are half-bytes
+    let Some(mut current_node_ptr) = trie_storage.get_root() else {
+        return Ok(None);
+    };
+    loop {
+        match trie_storage.get_node(current_node_ptr, AccessOptions::DEFAULT)? {
+            GenericTrieNode::Empty => return Ok(None),
+            GenericTrieNode::Leaf { extension, .. } => {
+                key_nibbles.extend(extension);
+                return Ok(Some(NibbleSlice::nibbles_to_bytes(&key_nibbles)));
```

Comment:
Amusingly there are two methods called nibbles_to_bytes, the one you're using asserts that nibbles is of even length. Is it an invariant here that the leaves are always at trie keys with even length?

Reply:
I was unaware of this caveat. I can look for the other method or truncate it to be even.